source/SH3/graphics/texture.cpp

```diff
@@ -13,48 +13,330 @@
+};
+
+struct bgr
+{
+    std::uint8_t b, g, r;
+};
+
+struct tga_header
+{
+    std::uint8_t id_size;   // Size of the ID field that follows this header (usually 0)
+    std::uint8_t colormap;  // Is this image paletted (THIS MUST ALWAYS BE 0 FOR US!)
+    std::uint8_t type;      // Type of image (ALWAYS 2 FOR RGB24!)
+    std::uint8_t unused[5]; // Colormap crap
+
+    std::uint16_t x_origin;  // Co-ordinate for first X value
+    std::uint16_t y_origin;  // Co-ordinate for first Y value
+    std::uint16_t width;
+    std::uint16_t height;
+    std::uint8_t bpp;       // Bits per pixel
+    std::uint8_t flags;
+};
+
+#pragma pack(pop)
+
+/**
+ *  Dump a texture to a Microsoft BMP/DIB
+ *
+ *  @param width - The Width of this texture.
+ *  @param height - The Height of this texture.
+ *  @param data - Pixel data
+ */
+void Dump2Bitmap(std::uint32_t width, std::uint32_t height, std::vector<std::uint8_t>& data)
+{
+    tga_header header;
+    std::ofstream file("output.tga", std::ios::binary); // Open the stream for binary output
+
+    memset(&header, 0, sizeof(header));
+
+    header.id_size      = 0x00;
+    header.colormap     = 0x00;
+    header.type         = 2;
+
+    header.x_origin     = 0x0000;
+    header.y_origin     = 0x0000;
+    header.width        = (uint16_t)width;
+    header.height       = (uint16_t)height;
+    header.bpp          = 24;
+    header.flags        = 0x20; // Origin is upper left hand corner
+
+    file.write((char*)&header, sizeof(header));
+    file.write((char*)&data[0], data.size());
+}
+
 //TODO: Scale the texture and then
 void sh3_texture::Load(sh3_arc& mft, const std::string& filename)
 {
     sh3_texture_header          header;
     sh3_arc_vfile               file(mft, filename);
     sh3_arc_vfile::read_error   e;
-    std::vector<std::uint8_t>   data;     // Pixel data of this texture (with the header stripped)
+    std::vector<uint8_t>        data;       // Pixel data of this texture (with the header stripped)
+    std::vector<rgba>           palette;    // Palette Data (I think this is BGRA)
+    std::size_t                 ret;        // How many bytes we get in from file.ReadData( )
 
-    std::size_t ret = file.ReadData(&header, sizeof(header), e); // Read in the header
+    ret = file.ReadData(&header, sizeof(header), e);
 
     if(ret != sizeof(header))
-        die("sh3_texture::Load( ): Header only read %d bytes of data! It should have been %d", ret, sizeof(header)); // TODO: Fallback on dummy texture (like Source's purple and black checkboard)
+        die("sh3_texture::Load( ): ret != sizeof(header)!");
 
-    if(header.bpp == 8)
-        Log(LogLevel::WARN, "WARNING: 8-bit textures are currently unsupported! They WILL look like nothing but pixel garbage.");
+    // Check for the pesky 64-byte A7A7A7A7 header that sometimes precedes our texture header
+    if(header.batchHeaderMarker == 0x00000000 && header.batchSize == 0xA7A7A7A7) // AHA!
+    {
+        offset = HEADER_START;
```

Comment:
How do you mean? Just set offset to 0x40 without worrying about a constant and just comment it?